de.metas.contracts/src/main/java/de/metas/contracts/interceptor/C_Flatrate_Term.java

```diff
@@ -569,4 +585,115 @@ public void setMasterDocumentNo(final I_C_Flatrate_Term term)
 			}
 		}
 	}
+	
+	
+	@ModelChange(timings = {
+			ModelValidator.TYPE_BEFORE_NEW, ModelValidator.TYPE_AFTER_CHANGE
+	}, ifColumnsChanged = {
+			I_C_Flatrate_Term.COLUMNNAME_ContractStatus, I_C_Flatrate_Term.COLUMNNAME_C_FlatrateTerm_Next_ID
+	})
+	public void updateContractStatusInOrder(final I_C_Flatrate_Term term)
+	{
+		final ContractOrderRepository contractOrderRepository = Adempiere.getBean(ContractOrderRepository.class);
+		final ContractOrderService contractOrderService = Adempiere.getBean(ContractOrderService.class);
+		
+		final OrderId orderId = contractOrderService.getContractOrderId(term);
+		if (orderId == null)
+		{
+			return;
+		}
+		
+		if (InterfaceWrapperHelper.isNew(term) && !Services.get(IContractsDAO.class).termHasAPredecessor(term))
+		{
+			final I_C_Order contractOrder = Services.get(IOrderDAO.class).getById(orderId, I_C_Order.class);
+			contractOrderRepository.setOrderContractStatusAndSave(contractOrder, I_C_Order.CONTRACTSTATUS_Active);
+			return;
+		}
+
+		final Set<OrderId> orderIds = contractOrderService.retrieveAllContractOrderList(orderId);
+
+		updateStatusIfNeededWhenExtendind(term, orderIds);
+
+		updateStatusIfNeededWhenCancelling(term, orderIds);
+
+		updateStausIfNeededWhenVoiding(term);
+	}
+
+	private void updateStatusIfNeededWhenExtendind(final I_C_Flatrate_Term term, final Set<OrderId> orderIds)
+	{
+		if (InterfaceWrapperHelper.isValueChanged(term, I_C_Flatrate_Term.COLUMNNAME_C_FlatrateTerm_Next_ID)
+				&& term.getC_FlatrateTerm_Next_ID() > 0
+				&& orderIds.size() > 1) // we set the Extended status only when an order was extended
+		{
+			final ContractOrderRepository contractOrderRepository = Adempiere.getBean(ContractOrderRepository.class);
+			final ContractOrderService contractOrderService = Adempiere.getBean(ContractOrderService.class);
+			
+			// update order contract status to extended
+			final OrderId currentContractOrderId = contractOrderService.getContractOrderId(term);
+			
+			orderIds.forEach(id -> {
+				if (id.getRepoId() != currentContractOrderId.getRepoId())  // different order from the current one
+				{
+					final I_C_Order order = Services.get(IOrderDAO.class).getById(id, I_C_Order.class);
+					contractOrderRepository.setOrderContractStatusAndSave(order, I_C_Order.CONTRACTSTATUS_Extended);
+				}
+			});
+		}
+	}
+
+	private void updateStatusIfNeededWhenCancelling(final I_C_Flatrate_Term term, final Set<OrderId> orderIds)
+	{
+		if (X_C_Flatrate_Term.CONTRACTSTATUS_EndingContract.equals(term.getContractStatus())
+				|| X_C_Flatrate_Term.CONTRACTSTATUS_Quit.equals(term.getContractStatus()))
+		{
+			final ContractOrderRepository contractOrderRepository = Adempiere.getBean(ContractOrderRepository.class);
+			
+			// update order contract status to cancelled
+			orderIds.forEach(id -> {
```

Comment:
i don'T see a benefit over a classical java-5 for loop. it's just more cumbersome to debug imho